david-fraley/MassVaccinationSystem
# [MVS-633] Connect Patient Registration and Broker

## :floppy_disk: [Code Check]
_Make sure you've checked/done the following before making this PR!_
* Does your code build/run?
* Does your design approach make sense?
* Does your work achieve the intention of the card?
* Does your code have any unintentional side effects -- have you tested a full workflow & checked the logic branches you touched?
* Did you make a unit test for your code (as appropriate)?
* Is your branch relatively up to date with the Development branch (within ~2 commits)
* Have you moved the card to "Needs QA" in Jira


## :flipper: [JIRA Sprint Card]

https://mass-immunization-system.atlassian.net/browse/MVS-633

## :newspaper: [Work Description]

Registering a patient through the application will create a FHIR resource for the patient in the backend.
Broker returns a QR code that is unique to the patient.
Mocked appointment - created and linked an appointment resource to each patient.

## :vertical_traffic_light: [Testing/QA Notes]

Setup:
- Define these variables in root/.env. All variables are configurable.
> FHIR_URL_BASE=http://localhost:8080/hapi-fhir-jpaserver/fhir
> 
> BROKER_DB_NAME=massvaxx
> BROKER_DB_USERNAME=massvaxxadmin
> BROKER_DB_PASSWORD=BROKER_DB_PASSWORD
> 
> HAPI_DB_NAME=hapi
> HAPI_DB_USERNAME=hapiuser
> HAPI_DB_PASSWORD=HAPI_DB_PASSWORD
- and hapi_fhir/data/hapi/hapi.properties has matching configuration (lines 43-45).

Build:
- Details in Readme
- If you get this error in the docker logs, https://stackoverflow.com/questions/14219092/bash-script-and-bin-bashm-bad-interpreter-no-such-file-or-directory, go to broker_db/ and run `vim massvaxx_db_init.sh`. When the vim editor opens, type `:set ff=unix`, which will appear in the bottom left, followed by `:wq`.
- If you get an error connecting to the database, try running `docker-compose down -v` or `docker-compose -f docker-compose-dev.yml down -v` depending on build method (Note: FHIR resources will be deleted). If issue persists, try `docker system prune -af --volumes`

Testing:
- Go through Single and Household Registration workflow
- Verify QR code appears in browser console after submitting the record (may have to wait for a few seconds)
- Check the logs from docker and find the id of the patient created (look for a PUT request)
- Run PatientIdServiceUnitTest and EndpointsUnitTest, in broker/testing/
